### DISCUSSION TITLE
Resource reachability or do not optimize flag

### DISCUSSION THREAD

**Discussion by BLCK-B** [oracle/graal#9479] [Q&A] [answered]
Hi, in a Java project I'm trying to prevent native build optimizing Vue.js files in `resources/static/`. I tried the JSON config according to [docs](https://www.graalvm.org/latest/reference-manual/native-image/metadata/#resource-metadata-in-json) unsuccessfully. Even if I adjust the js code, analysis does not see all valid paths.
What is the recommended approach to skip optimization of certain files, ideally all inside the dir?

**Comment by wirthi** [comment]
Can you please add some context here? Are you using Graal.js to run javascript code; what is your intent with those JavaScript files? What do you mean "optimizing Vue.js files" - including/excluding them in the built image?

Native-image will only look at Java code, not JavaScript files. So without more context, we can't really help you, we need to understand what you are up to, what works and what does not work.

**Comment by BLCK-B** [comment]
No, it is Java Graal. The Java Spring project serves static JavaScript frontend files located in resources/static/. Only the native executable throws errors in the browser. A bootJar or an IDE work fine.
The affected JavaScript file is identical in both bootJar and the native executable. Thus, I'm hesitant raising an issue yet.

I made a simple example.

IDE / bootJar:
<img src="https://github.com/user-attachments/assets/63bf3f76-c5a6-4f89-82cb-3f60cc7da986" width="480px"/> <br>
native exe:
<img src="https://github.com/user-attachments/assets/6250d569-52b6-4dd9-91cd-383b41a44f86" width="480px"/> <br>

to reproduce:
https://github.com/BLCK-B/Graal-issue

- from topmost dir `./gradlew bootRun`
this opens browser, program serves static Vue.js files: a table with data

- stop the process
- build native executable `./gradlew nativeCompile`
- run the executable from build/native/nativeCompile/
the table does not load data, look in browser console

GraalVM 22.0.2+9.1
Windows 11
setup: see [build.gradle](https://github.com/BLCK-B/Graal-issue/blob/main/build.gradle)

Do you have an idea where the difference could be? It is puzzling.

**Comment by wirthi** [comment, marked as answer]
Hi,

you are not providing any configuration for the ahead-of-time compilation. Native-Image needs to understand what e.g. resources you want to have included.

The tracing agent can provide that information based on a test run that covers the relevant usecases:
https://www.graalvm.org/latest/reference-manual/native-image/metadata/AutomaticMetadataCollection/
The output won't be perfect and will certainly need manual tuning. 

I got that working by adding this to the build.gradle:
```
bootRun {
    jvmArgs("-agentlib:native-image-agent=config-output-dir=tracing")
}
```

The content of the tracing dir I then copied to `build/resources/aot/META-INF/native-image` - there is probably a better way to do this in grade, but it did the trick for me. Deleting the NativeExample and building again worked for me then.

**Reply by BLCK-B** [reply]
This is interesting.  May I ask how you terminate the process? I tried all manners (IntelliJ IDEA), including ctrl+c and spring's `server.shutdown=graceful`, but every time the lock file remains and no data is written.
And secondly, while this is the correct approach, isn't there an explicit flag that would tell the Native-Image to include a specific folder? I tried `-H:IncludeResources` to no avail.

**Reply by wirthi** [reply]
`-H:IncludeResources="<Java regexp that matches resources to be included in the executable>"` should work. Use `-H:Log=registerResource:<log level>` (e.g. with level `3`) to debug what files are actually included. 

Reference Manual: https://www.graalvm.org/latest/reference-manual/native-image/dynamic-features/Resources/

**Reply by BLCK-B** [reply]
Thanks for the clear descriptions. I tried all sorts of combinations of IncludeResources. Then I managed to generate tracing from jar at least. I now have sufficient information to continue.
